throw errors when invalid values passed to dash(...)

floyd-may opened the pull request:
<!--
Please make sure you are familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- Is it a Bug fix, feature, docs update, ... -->

**The dash(...) function now throws errors when passed invalid values**

<!-- You can also link to an open issue here -->
Fixes #949. Current behavior is that `dash(...)` accepts values such as `0` for length, which breaks rendering in some viewers. There is a demo in the linked issue of the impact of the problem.

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [x] Tests (preference for unit tests)
- [x] Documentation
- [x] Update CHANGELOG.md
- [x] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->

<!-- feel free to add additional comments -->

<!-- Thank you for contributing! -->

floyd-may commented:
@danthegoodman feedback appreciated 😄

danthegoodman commented on lib/mixins/vector.js:
```diff
@@ -63,14 +63,18 @@ export default {
       return this.addContent(`[${length}] ${number(phase)} d`);
     } else {
+      if (!Array.isArray(length) && !Number.isInteger(length) || length === 0) {
+        throw new Error(`dash(${length}) invalid, must be number greater than zero`);
```
1. `!Array.isArray(length)` has already been checked and will always return true. I think it should be omitted here.
2. The length doesn't have to be an integer. The PDF spec allows floating point numbers. `Number.isFinite(length)` should get you your null check while also covering cases like `NaN` and strings and infinities.
3. Your check is only for zero. You should also check for negative numbers.

floyd-may replied:
Agreed, nice catches. I went ahead and consolidated the parallel paths in the `if` and `else` - seems cleaner now.

danthegoodman commented on lib/mixins/vector.js:
```diff
@@ -63,14 +63,18 @@ export default {
-      return this;
-    }
     if (Array.isArray(length)) {
+      const hasZeros = length.filter(x => x === 0).length > 0;
```
1. Using either the `every` or `some` function here will be a little more efficient, since they can short-circuit and doesn't build up an intermediate array. 
2. Like the non-array check, this check should reject negative numbers, `NaN`, nulls and other invalid values.

floyd-may replied:
👍

floyd-may commented:
@danthegoodman I think it's ready for another look.

blikblum commented:
I'm afraid this will break rendering since does not handle space option.

Will do some tests.

danthegoodman commented:
@blikblum You're right. I missed that in my second review. Good catch.

floyd-may commented:
I'll get it squared in a couple days unless someone beats me to it.

(misspelled from my mobile phone)

On Fri, May 3, 2019, 10:51 PM Danny Kirchmeier <notifications@github.com>
wrote:

> @blikblum <https://github.com/blikblum> You're right. I missed that in my
> second review. Good catch.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/foliojs/pdfkit/pull/969#issuecomment-489290872>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMIYNA3WAOIOWHVIGHW3J3PTUB5FANCNFSM4HKUFARA>
> .
>

blikblum commented:
I added dash unit tests to ensure no regressions are made with https://github.com/foliojs/pdfkit/commit/f486d11efe25105bf423d0f6e197bfa4d06e8545

You should rebase and add the throw tests in `vector.spec.js`. As long as unit tests are running all is fine

floyd-may commented:
@blikblum thanks for those unit tests for `dash(...)` - that really illustrated how badly I had broken things 😅

I moved the new tests to `vector.spec.js` and fixed the issues that your unit tests revealed. I went ahead and rewrote the PR as a single commit as well.

blikblum commented on lib/mixins/vector.js:
```diff
@@ -63,20 +63,18 @@ export default {
     }
+    length = length.map(v => number(v)).join(' ');
+    phase = options.phase || 0;
+    return this.addContent(`[${length}] ${number(phase)} d`);
```
```suggestion
    return this.addContent(`[${length}] ${number(options.phase || 0)} d`);
```
And remove phase variable

floyd-may commented:
@blikblum one more look please?

blikblum commented:
Many thanks

floyd-may commented:
Any guidance on when this fix will be available in a published release?
